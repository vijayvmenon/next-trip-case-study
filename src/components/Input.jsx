/* eslint-disable */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../constants';
import { ReactComponent as SearchIcon } from '../assets/search-icon.svg';

const useStyles = createUseStyles({
  root: {
    position: 'relative',
    marginTop: 24,
  },
  input: {
    fontSize: 18,
    padding: '14px 10px 0px 5px',
    display: 'block',
    width: 300,
    color: colors.TEXT_COLOR,
    height: 48,
    boxSizing: 'border-box',
    border: '1px solid #757575',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
    },
    '&:focus ~ label': {
      top: 3,
      left: -5,
      fontSize: 14,
      color: '#4285f4',
      transform: 'scale(0.8)',
    },
    '&:valid ~ label': {
      top: 3,
      left: -5,
      fontSize: 14,
      transform: 'scale(0.8)',
    },
  },
  label: {
    textTransform: 'uppercase',
    color: '#999',
    fontSize: 18,
    fontWeight: 'normal',
    position: 'absolute',
    pointerEvents: 'none',
    left: 10,
    top: 13,
    transition: '0.2s ease all',
  },
  searchIcon: {
    cursor: 'pointer',
    height: '1.8rem',
    width: '1.8rem',
    position: 'absolute',
    right: 0,
    bottom: 9,
    right: 12,
    color: colors.TEXT_COLOR,
  },
});
export default function Input({
  onSearch,
  label,
  type = 'text',
  value,
  setInput,
}) {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <input
        type="text"
        className={classes.input}
        required
        value={value}
        onChange={(e) => {
          if (type === 'number') {
            const re = /^[0-9\b]+$/;
            // if value is not blank, then test the regex
            if (e.target.value === '' || re.test(e.target.value)) {
              setInput(e.target.value);
            }
          } else {
            setInput(e.target.value);
          }
        }}
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            onSearch(value, e);
          }
        }}
      />
      <label htmlFor={label} className={classes.label}>
        {label}
      </label>
      <SearchIcon
        onClick={() => onSearch(value)}
        className={classes.searchIcon}
      />
    </div>
  );
}

Input.propTypes = {
  onSearch: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  setInput: PropTypes.func.isRequired,
};
