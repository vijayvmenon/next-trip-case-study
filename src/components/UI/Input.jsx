import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import { colors } from '../../constants';
import { ReactComponent as SearchIcon } from '../../assets/search-icon.svg';
import { ReactComponent as CloseIcon } from '../../assets/close-icon-snackbar.svg';

const useStyles = createUseStyles({
  root: {
    position: 'relative',
    marginTop: 24,
    display: 'flex',
    '@media(max-width:640px)': {
      margin: '24px 3rem 0px 3rem',
    },
  },
  inputFocus: {
    outline: 'none',
    borderTopLeftRadius: 10,
    borderBottomLeftRadius: 10,
  },
  input: {
    fontSize: 18,
    borderTopLeftRadius: 10,
    borderBottomLeftRadius: 10,
    padding: '14px 10px 0px 8px',
    display: 'block',
    width: '100%',
    color: colors.TEXT_COLOR,
    height: 56,
    boxSizing: 'border-box',
    borderLeft: '1px solid #757575',
    borderTop: '1px solid #757575',
    borderBottom: '1px solid #757575',
    borderRight: 'none',
    '&:focus': {
      borderColor: colors.BACKGROUND_2,
      borderWidth: 2,
      extend: 'inputFocus',
    },
    '&:focus ~ label': {
      top: 4,
      left: 10,
      fontSize: '0.8rem',
      color: '#4285f4',
      // transform: 'scale(0.8)',
    },
    '&:valid ~ label': {
      top: 4,
      left: 10,
      fontSize: '0.8rem',
      // transform: 'scale(0.8)',
    },
  },
  label: {
    color: '#999',
    fontSize: 18,
    fontWeight: 'normal',
    position: 'absolute',
    pointerEvents: 'none',
    left: 10,
    top: 17,
    transition: '0.2s ease all',
  },
  searchIcon: {
    cursor: 'pointer',
    height: '1.8rem',
    width: '1.8rem',
    // position: 'absolute',
    // bottom: 12,
    // right: 12,
    color: colors.TEXT_COLOR,
    '&:hover': {
      color: colors.BACKGROUND_2,
      height: '2rem',
      width: '2rem',
      transition: '0.2s ease all',
    },
  },
  closeIcon: {
    cursor: 'pointer',
    height: '1.4rem',
    width: '1.4rem',
    // position: 'absolute',
    // bottom: 15,
    // right: 48,
    color: colors.TEXT_COLOR,
    '&:hover': {
      color: colors.BACKGROUND_2,
    },
  },
  inputAdornments: {
    borderRight: '1px solid #757575',
    borderTop: '1px solid #757575',
    borderBottom: '1px solid #757575',
    marginLeft: -2,
    display: 'flex',
    borderTopRightRadius: 10,
    borderBottomRightRadius: 10,
  },
  inputAdornmentsFocus: {
    borderRight: `2px solid ${colors.BACKGROUND_2}`,
    borderTop: `2px solid ${colors.BACKGROUND_2}`,
    borderBottom: `2px solid ${colors.BACKGROUND_2}`,
    marginLeft: -2,
    display: 'flex',
    borderTopRightRadius: 10,
    borderBottomRightRadius: 10,
    borderColor: colors.BACKGROUND_2,
  },
  removeButtonStyle: {
    background: 'none',
    color: 'inherit',
    border: 'none',
    padding: 0,
    font: 'inherit',
    cursor: 'pointer',
    outline: 'inherit',
    paddingRight: '0.8rem',
    paddingTop: '0.5rem',
  },
  searchButton: {
    extend: 'removeButtonStyle',
  },
  closeButton: {
    extend: 'removeButtonStyle',
  },
});
export default function Input({
  onSearch,
  label,
  inputType = 'text',
  value,
  setInput,
  focusOnLoad,
}) {
  const classes = useStyles();
  const inputRef = useRef(null);
  const [focused, setFocused] = React.useState(false);
  const onFocus = () => setFocused(true);
  const onBlur = () => setFocused(false);
  // focus input on load
  useEffect(() => {
    if (focusOnLoad && inputRef.current) {
      inputRef.current.focus();
    }
  }, [focusOnLoad]);

  return (
    <div className={classes.root}>
      <input
        ref={inputRef}
        data-testid="custom-input"
        onFocus={onFocus}
        onBlur={onBlur}
        type="text"
        className={classes.input}
        required
        value={value}
        onChange={(e) => {
          if (inputType === 'number') {
            const re = /^[0-9\b]+$/;
            // if value is not blank, then test the regex
            if (e.target.value === '' || re.test(e.target.value)) {
              setInput(e.target.value);
            }
          } else {
            setInput(e.target.value);
          }
        }}
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            onSearch(value, e);
          }
        }}
      />
      <label htmlFor={label} className={classes.label}>
        {label}
      </label>
      <div
        className={
          focused ? classes.inputAdornmentsFocus : classes.inputAdornments
        }
      >
        {value && (
          <button
            type="button"
            className={classes.closeButton}
            onClick={() => setInput('')}
          >
            <CloseIcon className={classes.closeIcon} />
          </button>
        )}
        <button
          type="button"
          className={classes.searchButton}
          onClick={() => onSearch(value)}
          data-testid="input-search-icon-button"
        >
          {' '}
          <SearchIcon className={classes.searchIcon} />
        </button>
      </div>
    </div>
  );
}

Input.defaultProps = {
  inputType: 'text',
  focusOnLoad: true,
  label: '',
};

Input.propTypes = {
  onSearch: PropTypes.func.isRequired,
  label: PropTypes.string,
  inputType: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  setInput: PropTypes.func.isRequired,
  focusOnLoad: PropTypes.bool,
};
