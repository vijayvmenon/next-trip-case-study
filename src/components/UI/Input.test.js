import { render, fireEvent, screen } from '@testing-library/react';
import Input from './Input';
import renderer from 'react-test-renderer';

const props = {
  onSearch: jest.fn(),
  label: 'test',
  inputType: 'text',
  value: 'test',
  setInput: jest.fn(),
  focusOnLoad: true,
};

describe('Component:Input.jsx', () => {
  test('Input.jsx renders correctly', () => {
    const input = renderer.create(<Input {...props} />).toJSON();
    expect(input).toMatchSnapshot();
  });

  test('Input onSearch prop is called on Click of Search Icon', () => {
    const { getByTestId } = render(<Input {...props} />);
    fireEvent.click(getByTestId('input-search-icon-button'));
    fireEvent.keyDown(getByTestId('input-search-icon-button'), {
      key: 'Enter',
      code: 'Enter',
    });
    expect(props.onSearch).toHaveBeenCalledTimes(1);
  });

  test('Input onSearch prop is called on clicking Enter', () => {
    const { getByTestId } = render(<Input {...props} />);
    fireEvent.keyDown(getByTestId('custom-input'), {
      key: 'Enter',
      code: 'Enter',
    });
    expect(props.onSearch).toHaveBeenCalledTimes(1);
  });
});
