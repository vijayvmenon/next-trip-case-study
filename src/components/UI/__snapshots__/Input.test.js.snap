// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:Input.jsx Input.jsx renders correctly 1`] = `
<div
  className="root-0-2-1"
>
  <input
    className="input-0-2-3"
    data-testid="custom-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    required={true}
    type="text"
    value="test"
  />
  <label
    className="label-0-2-4"
    htmlFor="test"
  >
    test
  </label>
  <div
    className="inputAdornments-0-2-7"
  >
    <button
      className="closeButton-0-2-11"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="closeIcon-0-2-6"
      >
        close-icon-snackbar.svg
      </svg>
    </button>
    <button
      className="searchButton-0-2-10"
      data-testid="input-search-icon-button"
      onClick={[Function]}
      type="button"
    >
       
      <svg
        className="searchIcon-0-2-5"
      >
        search-icon.svg
      </svg>
    </button>
  </div>
</div>
`;
